peps/pep-0667.rst

```diff
@@ -31,112 +30,221 @@ consistent regardless of threading or coroutines.
+``locals()`` builtin in :term:`optimized scopes <py3.13:optimized scope>`:
+
+* retain the historical behaviour of having each call to ``locals()`` on a given frame
+  update a single shared snapshot of the local variables
+* make ``locals()`` return write-through proxy instances (similar
+  to ``frame.f_locals``)
+* make ``locals()`` return genuinely independent snapshots so that
+  attempts to change the values of local variables via ``exec()``
+  would be *consistently* ignored rather than being accepted in some circumstances
+
+The last option was chosen as the one which could most easily be explained in the
+language reference, and memorised by users:
+
+* the ``locals()`` builtin gives an instantaneous snapshot of the local variables in
+  optimized scopes, and read/write access in other scopes; and
+* ``frame.f_locals`` gives read/write access to the local variables in all scopes,
+  including optimized scopes
+
+This approach allows the intent of a piece of code to be clearer than it would be if both
+APIs granted full read/write access in optimized scopes, even when write access wasn't
+needed or desired. For additional details on this design decision, refer to :pep:`558`,
+especially the :ref:`pep-558-motivation` section and :ref:`pep-558-exec-eval-impact`.
+
+This approach is not without its drawbacks, which are covered
+in the Backwards Compatibility section below.
+
 Specification
 =============
 
-Python
-------
+Python API
+----------
 
-``frame.f_locals`` will return a view object on the frame that
-implements the ``collections.abc.Mapping`` interface.
+The ``frame.f_locals`` attribute
+''''''''''''''''''''''''''''''''
 
-For module and class scopes ``frame.f_locals`` will be a dictionary,
-for function scopes it will be a custom class.
+For module and class scopes (including ``exec()`` and ``eval()``
+invocations), ``frame.f_locals`` is a direct
+reference to the local variable namespace used in code execution.
 
-``locals()`` will be defined as::
+For function scopes (and other :term:`optimized scopes <py3.13:optimized scope>`)
+it will be an instance of a new write-through proxy type that can directly modify
+the optimized local variable storage array in the underlying frame, as well as the
+contents of any cell references to non-local variables.
 
-    def locals():
-        frame = sys._getframe(1)
-        f_locals = frame.f_locals
-        if frame.is_function():
-            f_locals = dict(f_locals)
-        return f_locals
+The view objects fully implement the ``collections.abc.Mapping`` interface,
+and also implement the following mutable mapping operations:
+
+* using assignment to add new key/value pairs 
+* using assignment to update the value associated with a key
+* conditional assignment via the ``setdefault()`` method
+* bulk updates via the ``update()`` method
+
+Using ``del``, ``pop()`` or ``clear()`` to remove keys that correspond to local
+variables on the underlying frame is NOT supported.
```

Comment:
@markshannon The previous wording here looks like it was correct, as 3.13.0 isn't allowing even extra variables to be removed, and `.pop()` and `.clear()` don't exist on the frame proxy instances:

```
>>> import sys
... def g():
...     a = 1
...     yield sys._getframe().f_locals
...
>>> gen = g()
>>> live_locals = next(gen)
>>> live_locals
{'a': 1}
>>> live_locals["b"] = 2
>>> live_locals
{'a': 1, 'b': 2}
>>> del live_locals["b"]
Traceback (most recent call last):
  File "<python-input-8>", line 1, in <module>
    del live_locals["b"]
        ~~~~~~~~~~~^^^^^
TypeError: cannot remove variables from FrameLocalsProxy
>>> hasattr(live_locals, "pop")
False
>>> hasattr(live_locals, "clear")
False
```

I'll reword this again to *allow* implementations to support deleting extra variables (so we can change it in CPython, as I agree the status quo seems overly restrictive).

Reply:
Tangent: I encountered https://github.com/python/cpython/issues/125723 while checking this behaviour.

Reply:
Updated wording has been pushed. I'm not fond of declaring something to still be implementation defined, but compared to the previous situation, restricting that to only `del f_locals["extra_var"]`, `f_locals.pop("extra_var")`, and what happens when calling `f_locals.clear()` is still a big step forward (and I don't think we want to commit to the current state of CPython in this area as being the way that things *should* be).

Reply:
OK, my confusion above was due to assuming that this was something that had already been changed, when it's actually something that was reported as a bug in 3.13.0 and is in the process of being fixed (see https://github.com/python/cpython/issues/125590).

So, what this bit actually needs to say:

* `.clear()` is outright not supported (since it isn't clear what it should do with local variables, omitting it forces the consuming code to make its own choices)
* `del` and `.pop()` *are* supported, but raise `ValueError` for actual local variable keys

Reply:
OK, I think what I just pushed covers this correctly now (I also added links to the bug report and associated PR in the implementation notes section)